src/Microsoft.AspNetCore.SystemWebAdapters/HttpServerUtility.cs

```diff
@@ -17,8 +19,19 @@ internal HttpServerUtility(HttpContextCore context)
 
     public string MachineName => Environment.MachineName;
 
-    [Obsolete(Constants.NotImplemented)]
-    public string MapPath(string path) => throw new NotImplementedException();
+    public string MapPath(string? path)
+    {
+        var appPath = (string.IsNullOrEmpty(path) ? VirtualPathUtility.GetDirectory(_context.Request.Path) :
+            VirtualPathUtility.Combine(
+                VirtualPathUtility.GetDirectory(_context.Request.Path) ?? "/"
+                , path));
+        // Potential alternative implementations here - for the most literal match to the original use HttpRuntime.AppDomainAppPath
```

Comment:
How should we decide the approach here? What are the pros and cons? I'm wondering if this is something we'd want to provide options to toggle between (either globally or per endpoint).

Reply:
I was thinking maybe an extension method/overload for `MapPath` that allows passing the root path, (or an enum value) to say which scenario you want. This would allow you to use your existing code as-is, but use the overload on a per-use case scenario. If it was available to both Core and Framework, you could adjust your code as necessary and still have it work on both sides (e.g. from a class library).

I could also see adding some kind of IMapPathRoot to the services where you could handle it as you wish. This could also maybe match up with something from StaticFiles / PhysicalFileProvider / CompositeFileProvide if you are mapping multiple folders?

I'm really not sure the best approach here - and the current implementation follows the original System.Web approach of mapping directly to the 'website root' just now.

I guess a user could also create their own `IHttpRuntime` with the valid path for their use case (although maybe in that case we should make `IISHttpRuntime` public and AppDomainAppPath virtual? (Kind of like the tests do just now)

Reply:
I'm not sure how well it would work, but here is (yet) another implementation:
```cs
        var appPath = (string.IsNullOrEmpty(path) ? VirtualPathUtility.GetDirectory(_context.Request.Path) :
            VirtualPathUtility.Combine(
                VirtualPathUtility.GetDirectory(_context.Request.Path) ?? "/"
                , path));
        var rootProvider = _context.RequestServices.GetRequiredService<IWebHostEnvironment>().WebRootFileProvider;
        return rootProvider.GetFileInfo(appPath).PhysicalPath;
```
This might work for the various scenarios listed in https://docs.microsoft.com/en-us/aspnet/core/fundamentals/static-files?view=aspnetcore-6.0 like [Serve files from multiple locations](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/static-files?view=aspnetcore-6.0#serve-files-from-multiple-locations).

Reply:
Hmm, yeah there's definitely a number of directions we can go here.  For now, we can just go with the content root as the default (that matches the application path in HttpRuntime).

We'll need to consider how to expose a way to override it through an API review. We're locking down the APIs for v1, so we'll probably wait for any non-System.Web APIs (i.e. things like IMapPath) to be enabled after we get v1 shipped.

If the content path is sufficient for now, we can move forward with the PR. Otherwise, we may want to hold off until after v1.

Reply:
I think for initial use, the implementation that is in the code just now is fine. There are use cases beyond the basics that System.Web supports natively (which is why I have IMapPath in my projects anyway), that could be implemented, but I don't think it is _required_ for this to be usable.

Reply:
I've placed some initial work for `IMapPath` and the different implementations at https://github.com/CZEMacLeod/systemweb-adapters/tree/czemacleod/imappath
This is loosely based on the discussions above, combined with some Framework implementation I already have in production.
Feel free to see if any of it would be of interest - although as you say, this is probably a post V1 thing.